File: ios/RCTIrPayerManager.m
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>
#import "React/RCTEventEmitter.h"
#import "React/RCTUIManager.h"
#import <React/RCTLog.h>

@interface RCT_EXTERN_MODULE(RCTIrPlayerManager, RCTViewManager)


#pragma mark Extern-Methods
RCT_EXTERN_METHOD(play:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(pause:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(togglePlay:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(stop:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(volumeUp:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(volumeDown:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(jumpForward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(jumpBackward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(longJumpForward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(longJumpBackward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(shortJumpForward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(shortJumpBackward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(mediumJumpForward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(mediumJumpBackward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(extraShortJumpForward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(extraShortJumpBackward:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(getMediaInfo:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(rewind:(nonnull NSNumber *)node)
RCT_EXTERN_METHOD(fastForward:(nonnull NSNumber *)node)


#pragma mark Extern-Methods-With-Parameters
RCT_EXTERN_METHOD(setMediaTime:(nonnull NSNumber *)node time:(nonnull NSNumber *)time)
RCT_EXTERN_METHOD(setVolume:(nonnull NSNumber *)node volume:(nonnull NSNumber *)volume)
RCT_EXTERN_METHOD(addPlaybackSubtitle:(nonnull NSNumber *)node subTitleSrc:(nonnull NSString *)subTitleSrc)
RCT_EXTERN_METHOD(addPlaybackAudio:(nonnull NSNumber *)node subTitleSrc:(nonnull NSString *)audioSrc)
RCT_EXTERN_METHOD(setCurrentVideoSubTitleIndex:(nonnull NSNumber *)node index:(nonnull NSNumber *)index)


#pragma mark Extern-Methods-With-Callback


#pragma mark Properties
RCT_EXPORT_VIEW_PROPERTY(src, NSString);
RCT_EXPORT_VIEW_PROPERTY(width, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(height, NSNumber)

#pragma mark Properties-With-CallBack
RCT_EXPORT_VIEW_PROPERTY(onStopped, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onBuffering, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onEnded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onError, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onEsAdded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onOpening, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPaused, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPlaying, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onStateChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onTimeChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onGetMediaInfo, RCTDirectEventBlock)

#pragma mark Custom-Properties



















//RCT_CUSTOM_VIEW_PROPERTY(region, MKCoordinateRegion, MKMapView)
//{
//    [view setRegion:json ? [RCTConvert MKCoordinateRegion:json] : defaultView.region animated:YES];
//}


//    RCT_EXPORT_VIEW_PROPERTY(onPlay, RCTBubblingEventBlock)
//    RCT_EXPORT_VIEW_PROPERTY(onPause, RCTDirectEventBlock)
//    RCT_EXPORT_VIEW_PROPERTY(onPlay, RCTDirectEventBlock)

//RCT_EXTERN_METHOD(getCount: (RCTResponseSenderBlock)callback)

//RCT_EXTERN_METHOD(
//  decrement: (RCTPromiseResolveBlock)resolve
//  rejecter: (RCTPromiseRejectBlock)reject
//)

//_RCT_EXTERN_REMAP_METHOD(inc, increment, false)


//@interface RCT_EXTERN_MODULE(Counter, RCTEventEmitter)

@end

File: ios/RCTIrPayerManager.swift
//
//  IrPayerManager.swift
//  IrPlayer
//
//  Created by Maziar on 10/15/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation


@objc (RCTIrPlayerManager)
class RCTIrPlayerManager: RCTViewManager {

    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func view() -> UIView! {
        let view = IrPlayer(frame: CGRect(x: 0, y: 0, width: 100, height: 100))
        return view
    }
    
    @objc func play(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.play()
        }
    }

    @objc func pause(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.pause()
        }
    }
    
    @objc func togglePlay(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.togglePlay()
        }
    }
    
    @objc func stop(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.stop()
        }
    }
    
    @objc func setMediaTime(_ node: NSNumber, time: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.setMediaTime(time: time)
        }
    }
    
    @objc func volumeUp(_ node: NSNumber, volume: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.volumeUp()
        }
    }
    
    @objc func volumeDown(_ node: NSNumber, volume: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.volumeDown()
        }
    }
    
    @objc func setVolume(_ node: NSNumber, volume: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.setVolume(volume: volume)
        }
    }
    
    
    @objc func jumpForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.jumpForward()
        }
    }
    
    @objc func jumpBackward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.jumpBackward()
        }
    }
    
    @objc func longJumpForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.longJumpForward()
        }
    }
    
    @objc func longJumpBackward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.longJumpBackward()
        }
    }
    
    @objc func shortJumpForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.shortJumpForward()
        }
    }
    
    @objc func shortJumpBackward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.shortJumpBackward()
        }
    }
    
    @objc func mediumJumpForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.mediumJumpForward()
        }
    }
    
    @objc func mediumJumpBackward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.mediumJumpBackward()
        }
    }
    
    @objc func extraShortJumpForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.extraShortJumpForward()
        }
    }
    
    @objc func extraShortJumpBackward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.extraShortJumpBackward()
        }
    }
    
    @objc func addPlaybackSubtitle(_ node: NSNumber, subTitleSrc: NSString) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.addPlaybackSlaveSubtitle(subTitleSrc: subTitleSrc)
        }
    }
    
    @objc func addPlaybackAudio(_ node: NSNumber, audioSrc: NSString) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.addPlaybackSlaveAudio(audioSrc: audioSrc)
        }
    }
    
    @objc func fastForward(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.fastForward()
        }
    }
    
    @objc func rewind(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.rewind()
        }
    }
    
    @objc func getMediaInfo(_ node: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.getMediaInfo()
        }
    }
    
    @objc func setCurrentVideoSubTitleIndex(_ node: NSNumber, index: NSNumber) {
        DispatchQueue.main.async {
            let component = self.bridge.uiManager.view(
                forReactTag: node
            ) as! IrPlayer
            component.setCurrentVideoSubTitleIndex(index: index)
        }
    }
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
//    @objc
//    func getCount(_ callback: RCTResponseSenderBlock) {
//      callback([count])
//    }

//    @objc
//      func decrement(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) -> Void {
//        var count = 0
//        if (count == 0) {
//          let error = NSError(domain: "", code: 200, userInfo: nil)
//          reject("E_COUNT", "count cannot be negative", error)
//        } else {
//          count -= 1
//          resolve("count was decremented")
//        }
//      }
}
